File: src/web/app/api/chat/tools/execCommand.ts
import { tool } from 'ai'
import { z } from 'zod'
import { exec } from 'node:child_process'
import { promisify } from 'node:util'
import path from 'node:path'

const execAsync = promisify(exec)

// 命令在项目根目录下执行，与 readFile 保持一致
const PROJECT_ROOT = path.resolve(process.cwd(), '../..')

export const execCommandTool = tool({
  description: '在项目根目录下执行命令行命令，返回 stdout 和 stderr。适合运行脚本、查看目录结构、执行构建命令等操作。',
  inputSchema: z.object({
    command: z.string().describe('要执行的 shell 命令，例如 ls skills/ 或 python3 skills/get_user_db/scripts/query_user.py --id 1'),
    timeout_ms: z.number().optional().describe('超时时间（毫秒），默认 10000ms'),
  }),
  execute: async ({ command, timeout_ms = 10000 }) => {
    try {
      const { stdout, stderr } = await execAsync(command, {
        cwd: PROJECT_ROOT,
        timeout: timeout_ms,
        // 限制输出大小，避免超大输出撑爆 context
        maxBuffer: 1024 * 1024, // 1MB
      })
      return {
        success: true,
        command,
        stdout: stdout.trim(),
        stderr: stderr.trim(),
      }
    } catch (err: unknown) {
      const e = err as { message?: string; stdout?: string; stderr?: string }
      return {
        success: false,
        command,
        error: e.message ?? String(err),
        stdout: e.stdout?.trim() ?? '',
        stderr: e.stderr?.trim() ?? '',
      }
    }
  },
})

File: packages/server/app/tools/exec_command.py
"""exec_command 工具：在项目根目录执行 shell 命令。"""
from __future__ import annotations

import os
import subprocess

from pydantic import BaseModel, Field

from app.config import PROJECT_ROOT, settings
from app.models import ExecCommandResponse
from app.tools.base import ToolDef


class ExecCommandInput(BaseModel):
    command: str = Field(
        description="要执行的 shell 命令，例如 ls skills/ 或 python3 scripts/query.py --id 1"
    )
    timeout_ms: int = Field(
        default=10_000,
        ge=100,
        le=60_000,
        description="超时时间（毫秒），默认 10000ms，范围 100-60000ms",
    )


def _build_env() -> dict:
    """把 config.yaml 中的工具类凭证注入到子进程环境变量。

    为什么在这里注入而不是让脚本自己读 config.yaml？
      脚本可能是任意语言（Node.js、Python、Shell），
      统一由 Python 侧注入环境变量是最通用的方案，脚本只需读 process.env 即可。
    """
    env = os.environ.copy()
    t = settings.tools
    if t.qweather.api_key:
        env["QWEATHER_API_KEY"] = t.qweather.api_key
    if t.qweather.api_host:
        env["QWEATHER_API_HOST"] = t.qweather.api_host
    if t.serper.api_key:
        env["SERPER_API_KEY"] = t.serper.api_key
    if t.cloud_mail.token:
        env["CLOUD_MAIL_TOKEN"] = t.cloud_mail.token
    return env


def _execute(input: ExecCommandInput) -> ExecCommandResponse:
    try:
        completed = subprocess.run(
            input.command,
            cwd=PROJECT_ROOT,
            shell=True,
            capture_output=True,
            text=True,
            timeout=input.timeout_ms / 1000,
            check=False,
            env=_build_env(),
        )
        return ExecCommandResponse(
            success=completed.returncode == 0,
            command=input.command,
            stdout=completed.stdout.strip(),
            stderr=completed.stderr.strip(),
            error=None if completed.returncode == 0 else f"Command exited with code {completed.returncode}",
        )
    except subprocess.TimeoutExpired as exc:
        return ExecCommandResponse(
            success=False,
            command=input.command,
            stdout=(exc.stdout or "").strip(),
            stderr=(exc.stderr or "").strip(),
            error=f"Command timed out after {input.timeout_ms}ms",
        )


EXEC_COMMAND_TOOL = ToolDef(
    name="exec_command",
    description=(
        "在项目根目录下执行命令行命令，返回 stdout 和 stderr。"
        "适合运行脚本、查看目录结构、执行构建命令等操作。"
    ),
    input_model=ExecCommandInput,
    execute=_execute,
    is_read_only=False,
    is_destructive=False,
)
